vendor/zkllvm-metacraft-circuits/src/circuits_impl/weigh_justification_and_finalization.h

```diff
@@ -0,0 +1,193 @@
+        auto start_block_roots_gindex = DEPTH18_START_BLOCK_ROOTS_GINDEX;
+        auto index_in_block_roots = compute_start_slot_at_epoch_in_block_roots(epoch);
+        auto gindex = start_block_roots_gindex + index_in_block_roots;
+        ssz_verify_proof(beacon_state_root, block_root, proof, gindex);
+    }
+
+    void verify_finalized_checkpoint(Root beacon_state_root,
+                                     CheckpointVariable finalized_checkpoint,
+                                     BeaconStateLeafProof proof) {
+        auto finalized_checkpoint_leaf = hash_tree_root(finalized_checkpoint);
+        auto gindex = BEACON_STATE_FINALIZED_CHECKPOINT_GINDEX;
+        ssz_verify_proof(beacon_state_root, finalized_checkpoint_leaf, proof, gindex);
+    }
+
+    void assert_slot_is_not_first_in_epoch(Slot slot) {
+        auto slot_in_epoch = slot % SLOTS_PER_EPOCH;
+        assert_true(slot_in_epoch != 0);
+    }
+
+    bool is_supermajority_link(Gwei target_balance, Gwei total_active_balance) {
+        return target_balance * 3 >= total_active_balance * 2;
+    }
+
+    void process_justifications(const Gwei& total_active_balance,
+                                const Gwei& previous_epoch_target_balance,
+                                const Gwei& current_epoch_target_balance,
+                                const JustificationBitsVariable& justification_bits,
+                                const CheckpointVariable& current_justified_checkpoint,
+                                const CheckpointVariable& previous_epoch_checkpoint,
+                                const CheckpointVariable& current_epoch_checkpoint,
+                                // Outputs:
+                                CheckpointVariable& out_checkpoint,
+                                JustificationBitsVariable& out_justification_bits) {
+        const auto previous_epoch_supermajority_link_pred =
+            is_supermajority_link(previous_epoch_target_balance, total_active_balance);
+        const auto current_epoch_supermajority_link_pred =
+            is_supermajority_link(current_epoch_target_balance, total_active_balance);
+
+        auto new_current_justified_checkpoint =
+            previous_epoch_supermajority_link_pred ? previous_epoch_checkpoint : current_justified_checkpoint;
+
+        new_current_justified_checkpoint =
+            current_epoch_supermajority_link_pred ? current_epoch_checkpoint : new_current_justified_checkpoint;
+
+        const auto new_second_justification_bit =
+            previous_epoch_supermajority_link_pred ? true : justification_bits.bits[0];
+
+        auto new_justification_bits = justification_bits;
+        new_justification_bits.shift_right(1);
+        new_justification_bits.bits[1] = new_second_justification_bit;
+        new_justification_bits.bits[0] = current_epoch_supermajority_link_pred;
+
+        out_checkpoint = new_current_justified_checkpoint;
+        out_justification_bits = new_justification_bits;
+    }
```

Comment:
I really think you can copy this function almost verbatim from the consensus spec :D